Improve credit card naming and add Dashlane tests

SaintPatrck opened the pull request:
## 🎟️ Tracking

TBD

## 📔 Objective

This change introduces several improvements to the CXF (Credential Exchange Format) importer, focusing on better handling of credit card items and expanding test coverage.

The primary goal is to enhance the user experience when importing credit card data from CXF files by providing more sensible default names for items. A secondary objective is to add comprehensive integration tests for Dashlane exports to ensure compatibility and prevent regressions.

### Behavioral Changes:

When importing a credit card from a CXF file, if the item's main title is empty or consists only of whitespace, the importer will now use the cardholder's name as the item name. If both the title and cardholder name are absent, it will fall back to "Untitled Card". This prevents the creation of nameless credit card entries in the vault.

### Specific Changes:

- **`cxf/import.rs`**:
    - The internal `add_item` helper function has been updated to accept an optional `custom_name` parameter, allowing the item's name to be overridden during creation.
    - For credit card credentials, logic has been added to check if the `item.title` is blank. If it is, the cardholder's name (`card.cardholder_name`) is used as the item name. A fallback to "Untitled Card" is implemented if the cardholder's name is also unavailable.
    - Added several new unit tests to verify the new naming logic for credit cards under various conditions:
        - Empty title uses cardholder name.
        - Whitespace-only title uses cardholder name.
        - Title is present, so cardholder name is ignored for naming.
        - Both title and cardholder name are missing, resulting in a fallback name.

- **`cxf/tests/dashlane_import_test.rs`**:
    - A new integration test file has been added to validate the import of Dashlane-generated CXF files.
    - This includes a full test suite (`test_import_dashlane`) that parses a sample `dashlane_export.json` file and asserts that all items (Logins, Cards, Notes, Identities, etc.) are correctly converted into Bitwarden cipher objects.

- **`resources/dashlane_export.json`**:
    - A new resource file containing a sample export from Dashlane has been added. This file serves as the test case for the new Dashlane integration test.

- **`cxf/tests/mod.rs`**:
    - The new `dashlane_import_test` module is now included in the test suite.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

SaintPatrck committed:
Improve credit card naming and add Dashlane tests

This commit introduces several improvements to the CXF (Credential Exchange Format) importer, focusing on better handling of credit card items and expanding test coverage.

Objective:
The primary goal is to enhance the user experience when importing credit card data from CXF files by providing more sensible default names for items. A secondary objective is to add comprehensive integration tests for Dashlane exports to ensure compatibility and prevent regressions.

Behavioral Changes:
When importing a credit card from a CXF file, if the item's main title is empty or consists only of whitespace, the importer will now use the cardholder's name as the item name. If both the title and cardholder name are absent, it will fall back to "Untitled Card". This prevents the creation of nameless credit card entries in the vault.

Specific Changes:
- **`cxf/import.rs`**:
    - The internal `add_item` helper function has been updated to accept an optional `custom_name` parameter, allowing the item's name to be overridden during creation.
    - For credit card credentials, logic has been added to check if the `item.title` is blank. If it is, the cardholder's name (`card.cardholder_name`) is used as the item name. A fallback to "Untitled Card" is implemented if the cardholder's name is also unavailable.
    - Added several new unit tests to verify the new naming logic for credit cards under various conditions:
        - Empty title uses cardholder name.
        - Whitespace-only title uses cardholder name.
        - Title is present, so cardholder name is ignored for naming.
        - Both title and cardholder name are missing, resulting in a fallback name.

- **`cxf/tests/dashlane_import_test.rs`**:
    - A new integration test file has been added to validate the import of Dashlane-generated CXF files.
    - This includes a full test suite (`test_import_dashlane`) that parses a sample `dashlane_export.json` file and asserts that all items (Logins, Cards, Notes, Identities, etc.) are correctly converted into Bitwarden cipher objects.

- **`resources/dashlane_export.json`**:
    - A new resource file containing a sample export from Dashlane has been added. This file serves as the test case for the new Dashlane integration test.

- **`cxf/tests/mod.rs`**:
    - The new `dashlane_import_test` module is now included in the test suite.

coroiu commented:
CXF is owned by arch, adding @Hinton as reviewer. Ping me when he has approved it!

Hinton commented on crates/bitwarden-exporters/src/cxf/import.rs:
```diff
@@ -88,10 +88,11 @@ pub(super) fn parse_item(value: Item) -> Vec<ImportingCipher> {
 
     // Helper to add ciphers with consistent boilerplate
-    let mut add_item = |t: CipherType, fields: Vec<Field>| {
+    let mut add_item = |t: CipherType, fields: Vec<Field>, custom_name: Option<String>| {
```
I think we can improve  this slightly. What do you think about renaming `custom_name` to be `fallback_name` and if ` title` is empty we fallback to that value?

SaintPatrck committed:
Merge branch 'main' into phonkonen/cxf-dashlane-cc-import-improvements

SaintPatrck committed:
Refactor add_item to use fallback_name pattern

Rename custom_name parameter to fallback_name and change logic so that
the item title is preferred, with fallback_name only used when title is
empty. This simplifies the credit card handling by removing duplicate
empty-title checking.

Hinton commented on crates/bitwarden-exporters/src/cxf/import.rs:
```diff
@@ -88,10 +88,15 @@ pub(super) fn parse_item(value: Item) -> Vec<ImportingCipher> {
+            fallback_name.unwrap_or_else(|| value.title.clone())
+        } else {
+            value.title.clone()
+        };
```
It's a bit unfortunate that we need the unwrap_or_else and then clone it. Not sure if this is cleaner ...

```suggestion
        let name = match fallback_name {
            Some(fallback) if value.title.trim().is_empty() => fallback,
            _ => value.title.clone(),
        };
```

SaintPatrck replied:
Thanks for the suggestion. That's much more elegant. 👍

Hinton commented on crates/bitwarden-exporters/src/cxf/import.rs:
```diff
@@ -188,6 +207,8 @@ pub(super) fn parse_item(value: Item) -> Vec<ImportingCipher> {
         } else {
             // No ciphers created yet, create standalone custom fields secure note
+            // Note: Cannot use add_item here due to borrow checker - add_item closure
+            // holds mutable borrow of output, conflicting with first_mut() above
```
Is this comment relevant?

Hinton reviewed (commented):
Should we add fallback names to everything? We could make it required 🤔

SaintPatrck committed:
Cleanup fallback and remove comment

SaintPatrck commented:
Thanks @Hinton

I'm for adding `fallback_name` to all of these and making it required. I think that can be done in a separate PR once we come to a consensus on what desired fallback names should be for these other items.

SaintPatrck commented:
Thanks @djsmith85 and @harr1424!
